examples/mem-alloc-test/src/main.rs

```diff
@@ -0,0 +1,293 @@
+use rand::Rng;
+use num_cpus;
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread;
+use std::time::{Instant};
+
+#[derive(Debug, PartialEq)]
+enum MemSize {
+    Large,
+    Medium,
+    Small,
+}
+/* These 3 variables will store the average latency of allocation+access+deallocation
+ * per thread type i.e, if there are 2 small threads, 3 medium threads and
+ * 4 large threads each of which runs 10, 20 and 30 times, then
+ * avg_duration_small_thread will store average latency of 2*10=20 iterations,
+ * avg_duration_medium_thread will store average latency of 3*20=60 iterations,
+ * and avg_duration_large_thread will store average latency of 4*30=120 iterations
+ */
+struct Counters {
+    avg_duration_small_thread: f64,
+    avg_duration_medium_thread: f64,
+    avg_duration_large_thread: f64,
+    global_average: f64,
+}
+
+const PAGE_SIZE: usize = 4096;
+const TO_KB: usize = 1024;
+const TO_MB: usize = TO_KB * 1024;
+const TO_GB: usize = TO_MB * 1024;
+
+
+/* Set of configurable parameters. These will adjusted as necessary while
+ * recording the performance numbers
+ */
+const NUM_CPUS: usize = 2;
+
+const LIMIT_SMALL_THREAD: i32 = 2;
+const LIMIT_MEDIUM_THREAD: i32 = 2;
+const LIMIT_LARGE_THREAD: i32 = 2;
+
+const SCAN_INTERVAL_SMALL_THREAD: usize = 1 * TO_KB;
+const SCAN_INTERVAL_MEDIUM_THREAD: usize = 1 * TO_MB;
+const SCAN_INTERVAL_LARGE_THREAD: usize = 1 * TO_MB;
+
+const SMALL_THREAD_MEM_START: usize = 1;
+const SMALL_THREAD_MEM_END: usize = 512;
+const MEDIUM_THREAD_MEM_START: usize = 1;
+const MEDIUM_THREAD_MEM_END: usize = 2;
+const LARGE_THREAD_MEM_START: usize = 1;
+const LARGE_THREAD_MEM_END: usize = 2;
+
+const MAX_INDEX_CHECKS_PER_BUFFER: usize = 32;
+
+fn calculate_and_print_stat(
+    shared_mutex_clone: Arc<Mutex<Counters>>,
+    tid: i32,
+    memsize: &MemSize,
+    avg_thread_latency: f64,
+) {
+    /* TODO: Record some other statistical parameters like more detailed statistics
+     * than just average, such as standard deviation, minimum and maximum time,
+     * and p95/p99/p99.9 latency
+     */
+    //println!("thread {} took {}\n", tid, avg_thread_latency);
+    let mut data = shared_mutex_clone.lock().unwrap();
+    /* Please note this is an intermediate value. Once we get the sum of individual
+     * averages of all the threads, then we will divide it by the frequency of
+     * the corresponding thread memsize type.
+     */
+    match memsize {
+        MemSize::Large => {
+            data.avg_duration_large_thread += avg_thread_latency;
+        }
+        MemSize::Medium => {
+            data.avg_duration_medium_thread += avg_thread_latency;
+        }
+        MemSize::Small => {
+            data.avg_duration_small_thread += avg_thread_latency;
+        }
+    };
+}
+
+fn get_random_num(start: usize, end: usize) -> usize {
+    let mut rng = rand::thread_rng();
+    rng.gen_range(start..=end)
+}
+
+fn wait_per_thread(pair_clone: Arc<(Mutex<bool>, Condvar)>) {
```

Comment:
You can simply this by using a [Barrier](https://doc.rust-lang.org/std/sync/struct.Barrier.html)